### PR TITLE
Fix GitHub Actions and add AllenSDK Dockerfiles

### LINKED ISSUE #166: upgrade pynwb version
The current dockerfile has pynwb>=2.3.1 and the latest version is 2.8.1. The latest version contains fixes to critical issues, so it would be good to upgrade.

### PR DESCRIPTION
**Pull request by kabilar**
## Proposed changes

- [x] In #212 we added GitHub Actions to build all Docker images and push to Docker Hub.  The context was incorrect so only the `images/Dockerfile` was being built for all entries in [matrix strategy](https://github.com/dandi/dandi-hub/blob/b3065528239ac0f81328822b03abe6a3202ff784/.github/workflows/docker-push.yaml#L15-L25).  This pull request fixes this issue.
- [x] Revert rename of `Dockerfile` tag from `latest-base` back to `latest`
- [x] Remove AllenSDK packages from `Dockerfile` and `Dockerfile.gpu`
- [x] Create `Dockerfile.allensdk` and `Dockerfile.gpu.allensdk` with the AllenSDK packages.
  - To reduce the image sizes, remove `napari`, `brainrender`, `neuroglancer`, `nibabel`, and `nilearn`.  Further work can be done in the future to remove additional packages that are not needed for the AllenSDK images.
- [x] Upgrade pynwb version - Fix #166
- [x] Confirm that images build and push to my Docker Hub - [kabilar](https://hub.docker.com/r/kabilar/dandihub/tags).
  - [x] `Dockerfile`
  - [x] `Dockerfile.allensdk`
  - [x] `Dockerfile.gpu` 
  - [x] `Dockerfile.gpu.allensdk`
  - [ ] ~`Dockerfile.matlab`~
  - [ ] ~`Dockerfile.gpu.matlab`~
  - The `Dockerfile.matlab` and `Dockerfile.gpu.matlab` fail because the base image has changed.

### CODE REVIEW

**Comment by asmacdo** [comment]
Looks right as far as I can see. I guess we should look into larger build environments, but if AllenSDK isnt needed this seems like a reasonable workaround.

**Comment by satra** [comment]
just an fyi, the allen sdk is different from the openscope image that carter is building. it's fine to remove and we could create a separate allen sdk image if needed.

**Comment by vijayiyer05** [comment]
Thank you @kabilar for the update & framing up this request for a token. FYI to @apaezmw.
I will bring this to attention of relevant authorities.

**Comment by aranega** [comment]
Hi @kabilar , thanks a lot for the feedback! 

For the image building, token or any form of authentication shouldn't be required, I think the issue comes from the fact that the base images `jupyter-byoi-matlab-notebook` has been discontinued and is not available anymore in the `ghcr.io/mathworks-ref-arch`. 

We need to update the dockerfile to use the new version, which will require a little bit of work as it relies on a new base image. The good news in that is that there is chances that some packages that we were installing the current version of the dockerfile will not be required anymore, meaning less layers in the final image and, hopefully, a slimer final image requiring less space to build.

**Comment by kabilar** [comment]
Thank you @vijayiyer05 @aranega.  That sounds good.  I will remove the automated build and push of the `Dockerfile.matlab` and `Dockerfile.gpu.matlab` from this pull request.  Whenever the new Dockerfiles are ready, we can easily add them back to the GitHub Actions workflows.

**Comment by satra** [comment]
@kabilar - is the pointer accurate? it may be that the docker image has changed.

**Comment by kabilar** [comment]
> @kabilar - is the pointer accurate? it may be that the docker image has changed.

Hi @satra, are you referring to the base image: `ghcr.io/mathworks-ref-arch/matlab-integration-for-jupyter/jupyter-byoi-matlab-notebook`?  I'm not sure how to access it?  The only Docker images that I can find are on the [Mathworks Docker Hub](https://hub.docker.com/u/mathworks?page=1&search=), which are built from the [container-images](https://github.com/mathworks-ref-arch/container-images?tab=readme-ov-file) repository.

**Comment by kabilar** [comment]
> > @kabilar - is the pointer accurate? it may be that the docker image has changed.
> 
> Hi @satra, are you referring to the base image: `ghcr.io/mathworks-ref-arch/matlab-integration-for-jupyter/jupyter-byoi-matlab-notebook`? I'm not sure how to access it. The only Docker images that I can find are on the [Mathworks Docker Hub](https://hub.docker.com/u/mathworks?page=1&search=), which are built from the [container-images](https://github.com/mathworks-ref-arch/container-images?tab=readme-ov-file) repository.

Based on the [matlab-integration-for-jupyter](https://github.com/mathworks-ref-arch/matlab-integration-for-jupyter?tab=readme-ov-file#pre-built-images) readme, I should be able to pull from the GitHub Container Registry, but not sure whats going on.

**Comment by satra** [comment]
check the packages here:  https://github.com/orgs/mathworks-ref-arch/packages?repo_name=matlab-integration-for-jupyter

they seemed to have renamed the tags and did not update this dockerfile.

**Comment by kabilar** [comment]
> check the packages here: https://github.com/orgs/mathworks-ref-arch/packages?repo_name=matlab-integration-for-jupyter
> 
> they seemed to have renamed the tags and did not update this dockerfile.

Good catch, @satra.  Just test with the `jupyter-matlab-notebook` image.  It is able to pull the image but now fails downstream.  See [latest run](https://github.com/kabilar/dandi-hub/actions/runs/12322291731/job/34395544773).  Will defer to @aranega.  Going to proceed with remove the Matlab builds from the GA workflows.

**Comment by kabilar** [comment]
Hi @asmacdo @satra, please re-review when you have a chance.  The first comment is now updated with a description of the latest changes.

@asmacdo, once this is merged (and can be done next week), I will need you to update the dandihub deployment to implement these changes.

**Comment by kabilar** [comment]
Hi @asmacdo @satra, this pull request is ready for review and merge.  Once it is merged, it will automatically trigger a build and push of the Docker images.

**Review by satra — APPROVED** [review]

File: images/Dockerfile.gpu.allensdk
Comment (asmacdo) [inline images/Dockerfile.gpu.allensdk:1]: Would it make sense to use

`FROM dandiarchive/dandihub:latest-allensdk`

so we can reduce the duplication?

**Comment by satra** [comment]
do we have auto redeploys of the hub when this is merged? if not the allen-sdk changes won't go through.

**Comment by asmacdo** [comment]
> do we have auto redeploys of the hub when this is merged? if not the allen-sdk changes won't go through.

No, but I can redeploy this evening

File: images/Dockerfile.gpu.allensdk
Comment (asmacdo) [reply images/Dockerfile.gpu.allensdk:1]: We could also build many of these from dandihub:latest, but that would require a bit of a refactor, specifically, we'd want to make sure that latest is built (and pushed if being done in separate environments) prior to the others building.  IMO this could be regarded as a future improvement, not blocking merge.

Comment (kabilar) [reply images/Dockerfile.gpu.allensdk:1]: I think that the only issue will be that we tag everything as `latest` so if the jobs that build images start simultaneously then the GPU image may pull the older `latest-allensdk` image.

Comment (kabilar) [reply images/Dockerfile.gpu.allensdk:1]: Looks like had the same idea and hit `Enter` at the same time.  That sounds good.  We can implement this in the future.

**Pull request merged**
